# 1. From a pile of text to a table

**The question this notebook answers: what is one row?**

That sounds like a formality. It is not. Almost every mistake later in this course traces
back to a row meaning something different from what you assumed — a claim about *people*
tested on a table of *messages*, an average over a unit nobody cares about, a count that
silently double-counts.

So before any plot, two jobs:

1. **Get the data into rows that match the question.** Here that means a regular expression,
   because the raw material is a wall of text.
2. **Add what the question needs and the data does not have.** Enrichment. This is where
   most analyses are actually won, and it is the part people skip because it feels like
   admin.

We do both on a showcase corpus first, where you can check your work against something
known, and then on your own chat.

In [ ]:
import re

import pandas as pd
from goad_toolkit.datatransforms import Pipeline, TransformBase

from wa_analyzer.data import PROCESSED, load_own_chat, load_showcase

## 1.1 The showcase: five years of Ubuntu IRC

Two public IRC channels, `#ubuntu-uk` and `#ubuntu-nl`, from 2013 to 2017. Real conversation
between real people, with the same awkward shape your own chat export has.

It ships with this repo, so this runs offline.

In [2]:
irc = load_showcase("ubuntu_irc")
print(irc.shape)
irc.head()

(3371, 3)


,created,channel,text
0,2013-01-01,#ubuntu-uk,[00:00] <penguin42> Happy new year!\n[00:00] <...
1,2013-01-01,#ubuntu-nl,[10:41] <Guest23768> goede morgen\n[10:41] <NR...
2,2013-01-02,#ubuntu-nl,[11:41] <capthchaman> goeiendag iedereen\n[11:...
3,2013-01-02,#ubuntu-uk,[01:42] <ali1234> wtf another piece of sewing ...
4,2013-01-03,#ubuntu-uk,[00:03] <neuro> i can't believe what was just ...


Three columns and 3,371 rows. So one row is... what?

Look at the `text` column before assuming.

In [3]:
day = irc.iloc[900]
print(f"{day.channel}  {day.created.date()}\n")
print(day.text[:400])

#ubuntu-uk  2014-03-27

[00:00] <neuro> yay, it's officially pay day! (for me)
[00:08] <mapp> ;]
[00:08] <mapp> yay
[00:08] <mapp> lol thats so stupud how did that happen neuro
[03:00] <Azelphur> ali1234: hey, you alive?
[03:00] <ali1234> sure
[03:00] <Azelphur> ali1234: got a couple random questions about netbooting, figured you'd have all the answers
[03:01] <ali1234> maybe
[03:01] <Azelphur> 1) Can I set my router (op


**One row is a whole day of one channel**, with every message of that day packed into a
single string.

That is not a useful unit for anything we want to ask. "How long are messages?" "Who talks
most?" "When is the channel busy?" — every one of those needs *one row per message*.

Nothing is broken here. The data was simply stored for a different purpose than ours. This
is normal, and noticing it is the job.

## 1.2 Regular expressions: pulling structure out of text

Every line looks like this:

```
[00:08] <mapp> yay
```

Three things we want — a time, an author, a message — and some punctuation holding them
apart. A regular expression is a way of describing that pattern precisely enough for the
computer to pull the pieces out.

Let us build it against one line.

In [4]:
line = "[00:00] <neuro> yay, it's officially pay day! (for me)"

pattern = re.compile(r"^\[(\d{2}):(\d{2})\]\s+<(\S+)>\s+(.*)$")
pattern.match(line).groups()

('00', '00', 'neuro', "yay, it's officially pay day! (for me)")

Four groups back, in order. Reading the pattern piece by piece:

| piece | means |
|---|---|
| `^` | start of the line — so a `[` in the middle of a sentence cannot match |
| `\[` | a literal `[`. Backslashed, because a bare `[` starts a character class |
| `(\d{2})` | **group**: exactly two digits. `\d` is shorthand for `[0-9]` |
| `:` | a literal colon |
| `\]` | a literal `]` |
| `\s+` | one or more whitespace characters |
| `<(\S+)>` | angle brackets around a **group** of one or more *non*-whitespace characters |
| `(.*)` | **group**: any character, zero or more times — the message itself |
| `$` | end of the line |

The parentheses are what makes this useful. Without them the pattern only answers "does
this line look right?"; with them it hands back the four pieces.

Two habits worth taking from this:

- **Anchor with `^` and `$`** unless you have a reason not to. An unanchored pattern will
  cheerfully find a match in the middle of something you did not mean.
- **`\S+` beats `.*` when you know there is no whitespace.** `.*` is greedy: it takes as much
  as it can and only gives back when forced. For a nickname that is wrong, and the bug it
  causes is subtle.

Now run it over everything. One row per line, per day, per channel.

In [5]:
rows, unmatched = [], []
for day in irc.itertuples():
    for line in day.text.split("\n"):
        if not line.strip():
            continue
        m = pattern.match(line)
        if m:
            hh, mm, author, message = m.groups()
            rows.append((day.created, day.channel, int(hh), int(mm), author, message))
        else:
            unmatched.append((day.channel, line))

msgs = pd.DataFrame(rows, columns=["date", "channel", "hh", "mm", "author", "message"])
print(f"{len(msgs):,} messages, {len(unmatched):,} lines did not match")
msgs.head()

615,683 messages, 11,883 lines did not match


,date,channel,hh,mm,author,message
0,2013-01-01,#ubuntu-uk,0,0,penguin42,Happy new year!
1,2013-01-01,#ubuntu-uk,0,0,neuro,happy new year folks!
2,2013-01-01,#ubuntu-uk,0,0,daftykins,+1
3,2013-01-01,#ubuntu-uk,0,0,Azelphur,happy new year :p
4,2013-01-01,#ubuntu-uk,0,1,daftykins,"yes irssi, day changed indeed."


## 1.3 The part everyone skips: what did *not* match?

615,683 rows is a satisfying number and it is the wrong thing to look at. The interesting
number is the other one.

A regex that matches most lines feels like success. But the lines it missed are not random —
they are the ones that are *shaped differently*, which is exactly where the surprises live.
Never accept a parse without looking at its leftovers.

In [6]:
missed = pd.DataFrame(unmatched, columns=["channel", "line"])
print(f"{len(missed):,} unmatched lines ({len(missed) / (len(msgs) + len(missed)):.2%})\n")

# Do not just eyeball a sample -- put the leftovers into named buckets.
missed["kind"] = "something else"
missed.loc[missed.line.str.match(r"^\[\d{2}:\d{2}\]\s+\*\s"), "kind"] = "action line"
missed.loc[missed.line.str.contains(r"[\x00-\x08\x0b-\x1f]"), "kind"] = "control characters"
print(missed.kind.value_counts().to_string(), "\n")

for kind, group in missed.groupby("kind"):
    print(f"{kind}:")
    print(f"   {group.line.iloc[0][:88]!r}\n")

11,883 unmatched lines (1.89%)

kind
action line           11489
control characters      390
something else            4 

action line:
   '[00:26]  * dwatkins is in Tollcross, about a mile from the castle'

control characters:
   '\x01\x03 Alt Gr 2'

something else:
   '[21:19] <OerHeks>'



Two different things are hiding in there.

**Action lines.** IRC has a `/me` command, and it logs differently — `* nick waves` rather
than `<nick> waves`. Same information, different shape. Our pattern was written for one and
not the other.

**Control characters.** Lines starting with `\x01` and `\x03` are what is left of IRC's
colour and formatting codes, partly stripped when this corpus was built. The message text
survived; the `[HH:MM] <nick>` prefix did not.

The first is worth fixing — those are real messages by real people, and dropping them would
quietly bias anything we measure about who says what. The second is damage, and the honest
move is to count it and move on rather than pretend to recover it.

In [7]:
action = re.compile(r"^\[(\d{2}):(\d{2})\]\s+\*\s+(\S+)\s+(.*)$")

rows, still_missing = [], []
for day in irc.itertuples():
    for line in day.text.split("\n"):
        if not line.strip():
            continue
        m = pattern.match(line) or action.match(line)
        if m:
            hh, mm, author, message = m.groups()
            rows.append((day.created, day.channel, int(hh), int(mm), author, message,
                         bool(action.match(line))))
        else:
            still_missing.append(line)

msgs = pd.DataFrame(rows, columns=["date", "channel", "hh", "mm",
                                   "author", "message", "is_action"])
coverage = len(msgs) / (len(msgs) + len(still_missing))
print(f"{len(msgs):,} messages, {coverage:.2%} of lines parsed")
print(f"{msgs.is_action.sum():,} of them are action lines")

627,172 messages, 99.94% of lines parsed
11,489 of them are action lines


**Write that number down.** "We recovered 99.94% of lines; the remainder are corrupted
formatting codes" is a sentence that belongs in your report — it is the difference between
an analysis somebody can trust and one they cannot.

## 1.4 Enrichment, as a pipeline

We now have messages. The question needs things messages do not carry: what hour was this,
what day of the week, did it contain a link, how long was it.

You could write these as forty loose cells that have to be run in the right order, and
whose order you will not remember next week. Or you can make each step a named object and
chain them — which is what `goad_toolkit`'s `Pipeline` is for.

A step is a class with one method. That is the whole contract.

In [8]:
class TimeFeatures(TransformBase):
    """Derive calendar features from a timestamp column."""

    def transform(self, data: pd.DataFrame, column: str) -> pd.DataFrame:
        ts = pd.to_datetime(data[column])
        data["day_name"] = ts.dt.day_name()
        data["isoweek"] = ts.dt.isocalendar().week
        data["year_week"] = ts.dt.strftime("%Y-%W")
        return data

In [9]:
class RegexFeature(TransformBase):
    """Add a feature extracted from a text column with a regular expression."""

    def transform(
        self,
        data: pd.DataFrame,
        column: str,
        pattern: str,
        feature: str,
        mode: str = "count",
    ) -> pd.DataFrame:
        text = data[column].fillna("")
        if mode == "count":
            data[feature] = text.str.count(pattern)
        elif mode == "has":
            data[feature] = text.str.contains(pattern, regex=True)
        elif mode == "extract":
            data[feature] = text.str.extract(pattern, expand=False)
        else:
            raise ValueError(f"mode must be count/has/extract, got {mode!r}")
        return data

Two things about `RegexFeature` worth noticing, because both will bite you otherwise.

**The new column is called `feature`, not `name`.** `Pipeline.add(name=...)` already uses
`name` for the *step's* name, and the base class consumes it. Any transform that names an
output column has to call the parameter something else.

**Parameters are spelled out in the signature**, not swallowed by `**kwargs`. That is what
lets the base class check them, and it is what makes the step readable a month from now.

In [10]:
pipeline = Pipeline()
pipeline.add(TimeFeatures, column="date")
pipeline.add(RegexFeature, name="urls",
             column="message", pattern=r"https?://\S+", feature="has_url", mode="has")
pipeline.add(RegexFeature, name="questions",
             column="message", pattern=r"\?", feature="n_question", mode="count")
pipeline.add(RegexFeature, name="mentions",
             column="message", pattern=r"^(\S+)[:,]\s", feature="addressed_to", mode="extract")

enriched = pipeline.apply(msgs)
enriched.head()

Applying transforms:   0%|          | 0/4 [00:00<?, ?it/s]

Applying transforms:  25%|██▌       | 1/4 [00:00<00:02,  1.33it/s]

Applying transforms:  50%|█████     | 2/4 [00:00<00:00,  2.71it/s]

Applying transforms:  75%|███████▌  | 3/4 [00:00<00:00,  3.95it/s]

Applying transforms: 100%|██████████| 4/4 [00:01<00:00,  5.15it/s]

Applying transforms: 100%|██████████| 4/4 [00:01<00:00,  3.73it/s]

,date,channel,hh,mm,author,message,is_action,day_name,isoweek,year_week,has_url,n_question,addressed_to
0,2013-01-01,#ubuntu-uk,0,0,penguin42,Happy new year!,False,Tuesday,1,2013-00,False,0,NaN
1,2013-01-01,#ubuntu-uk,0,0,neuro,happy new year folks!,False,Tuesday,1,2013-00,False,0,NaN
2,2013-01-01,#ubuntu-uk,0,0,daftykins,+1,False,Tuesday,1,2013-00,False,0,NaN
3,2013-01-01,#ubuntu-uk,0,0,Azelphur,happy new year :p,False,Tuesday,1,2013-00,False,0,NaN
4,2013-01-01,#ubuntu-uk,0,1,daftykins,"yes irssi, day changed indeed.",False,Tuesday,1,2013-00,False,0,NaN


`addressed_to` is the one worth pausing on. On IRC people answer each other by name —
`daftykins: try rebooting`. That single regex turns a flat list of messages into something
with a *reply structure*, which is a different kind of data entirely. Lesson 7 builds a
social graph out of exactly this column.

That is what enrichment means: not tidying, but adding a dimension the raw data did not
have.

And because the pipeline is an object, it can tell you what it did.

In [11]:
print(pipeline)

Pipeline(
  TimeFeatures: TimeFeatures(column='date'),
  urls: RegexFeature(column='message', pattern='https?://\\S+', feature='has_url', mode='has'),
  questions: RegexFeature(column='message', pattern='\\?', feature='n_question', mode='count'),
  mentions: RegexFeature(column='message', pattern='^(\\S+)[:,]\\s', feature='addressed_to', mode='extract')
)


> **Your turn, briefly.** Add one more step of your own to the pipeline above. Some ideas:
> the number of capital letters, whether the message ends in a full stop, whether it
> contains a smiley, message length in words. Re-run `pipeline.apply(msgs)` and check the
> column appears.
>
> Keep the ones you find interesting — lesson 5 uses features exactly like these to work out
> what distinguishes one person's writing from another's.

## 1.5 "Cleaned" is a claim somebody else made

The publishers of this corpus say they removed bots. Worth checking before you build on top
of it, because a bot is not a person and every per-author statistic will quietly include it.

Try it **without reading any messages** — structure only. That habit generalises to spam and
abuse, where reading everything is not an option.

In [12]:
def author_profile(g: pd.DataFrame) -> pd.Series:
    return pd.Series({
        "n": len(g),
        "repeat_rate": 1 - g.message.nunique() / len(g),
        "active_days": g.date.nunique(),
        "busiest_day_share": g.groupby("date").size().max() / len(g),
        "median_length": g.message.str.len().median(),
    })


profiles = enriched.groupby("author").apply(author_profile, include_groups=False)
profiles = profiles[profiles.n >= 20]
print(f"{len(profiles):,} authors with at least 20 messages")
profiles.nlargest(8, "repeat_rate").round(3)

1,255 authors with at least 20 messages


,n,repeat_rate,active_days,busiest_day_share,median_length
author,,,,,
tiiiiiirnot,32.0,0.969,1.0,1.000,426.0
Adem,24.0,0.875,1.0,1.000,1.0
D_O_U_C_H_E,42.0,0.810,1.0,1.000,34.0
Mark_de_J,32.0,0.469,4.0,0.531,5.0
brobostigon,9161.0,0.416,1797.0,0.008,23.0
lubotu3`,127.0,0.386,58.0,0.173,85.0
hoover,70.0,0.329,34.0,0.143,11.0
NoirX,489.0,0.305,108.0,0.139,20.0


`repeat_rate` is the share of an author's messages that duplicate something they already
said. A script repeats itself constantly; a person, occasionally. So filter on it.

In [13]:
suspects = profiles[profiles.repeat_rate > 0.4].sort_values("repeat_rate", ascending=False)
suspects.round(3)

,n,repeat_rate,active_days,busiest_day_share,median_length
author,,,,,
tiiiiiirnot,32.0,0.969,1.0,1.000,426.0
Adem,24.0,0.875,1.0,1.000,1.0
D_O_U_C_H_E,42.0,0.810,1.0,1.000,34.0
Mark_de_J,32.0,0.469,4.0,0.531,5.0
brobostigon,9161.0,0.416,1797.0,0.008,23.0


The biggest of them is `brobostigon` — 9,161 messages over 1,797 days. Look at what they
repeated.

In [14]:
brobo = enriched[enriched.author == "brobostigon"]
top = brobo.message.value_counts().head(3)
for text, count in top.items():
    days = brobo[brobo.message == text].date.nunique()
    print(f"{count:>5}x  on {days:>5} different days   {text!r}")

 1335x  on  1334 different days   'morning boys and girls.'
  478x  on   353 different days   ':)'
  203x  on   203 different days   'good morning everyone,'


**"morning boys and girls."** 1,335 times, on 1,334 different days. Someone saying good
morning to their friends, every morning, for five years.

Now the bot.

In [15]:
bot = profiles.loc[["lubotu3`"]]
print(bot.round(3).to_string(), "\n")

lubotu = enriched[enriched.author == "lubotu3`"]
for text, count in lubotu.message.value_counts().head(3).items():
    print(f"{count:>5}x  {text!r}")

              n  repeat_rate  active_days  busiest_day_share  median_length
author                                                                     
lubotu3`  127.0        0.386         58.0              0.173           85.0 

   19x  "No, it's not out yet!"
    8x  'pong!'
    4x  'dong'


And `lubotu3\`` is the actual bot — it pastes canned answers when someone says a keyword.
Its repeat rate is 0.386, *below* the threshold that caught `brobostigon`.

Repetition does not separate scripts from people, because people have rituals.

> That greeting returns in lesson 5, where how someone writes stops being noise and becomes
> the thing being measured.

## 1.6 Your turn: your own chat

Everything above works the same on your own export. If you do not have one yet, this section
prints a message and the notebook still finishes — come back to it once you do.

The one extra step is anonymisation. `humanize` maps each real name to a stable nickname, so
you can hand in your work without handing over your friends' names.

In [ ]:
own = load_own_chat()

In [ ]:
if own is not None:
    from wa_analyzer.humanhasher import humanize

    anon = {name: humanize(name) for name in own.author.unique()}
    own["author"] = own.author.map(anon)
    print(f"{len(anon)} authors, anonymised. For example:")
    display(own.head(3))
else:
    print("No chat of your own yet — skipping. The showcase half above still stands.")

Now the *same pipeline*, pointed at a different dataframe. That is the payoff of having
written the steps as objects rather than as cells: nothing changes but the input.

In [ ]:
if own is not None:
    own_pipeline = Pipeline()
    own_pipeline.add(TimeFeatures, column="timestamp")
    own_pipeline.add(RegexFeature, name="urls",
                     column="message", pattern=r"https?://\S+",
                     feature="has_url", mode="has")
    own_pipeline.add(RegexFeature, name="questions",
                     column="message", pattern=r"\?",
                     feature="n_question", mode="count")

    own_enriched = own_pipeline.apply(own)
    display(own_enriched.head())

> **Your turn.** Write one `RegexFeature` step that is specific to *your* chat and would make
> no sense on IRC. Emoji, a group in-joke, a language you switch into, the way one person
> always signs off. Two of them.
>
> Chat data is unusually rich for this: a WhatsApp export gives you timestamps with daily and
> weekly rhythms, natural language, and half a dozen distributions with textbook shapes. Most
> of what makes this course work comes from features you extract yourself in this notebook.

In [ ]:
if own is not None:
    from datetime import datetime

    outfile = PROCESSED / f"chat-{datetime.now():%Y%m%d-%H%M%S}.parq"
    own_enriched.to_parquet(outfile, index=False)
    print(f"Wrote {outfile}")
    print("Put that filename after `current` in config.toml so the other notebooks find it.")

## 1.7 What to write down

Before moving on, you should be able to answer these about your own data. Not in your head —
written down, because lesson 2 starts by assuming them.

1. **What is one row, now?** One message. Say what a message is in your export — does a
   photo count, a system notice, a message that was deleted?
2. **How many rows did you lose, and to what?** Every parse drops something. Name the number
   and the reason.
3. **How many *people* are in your data?** Not messages — people. That number is much
   smaller, and in lesson 2 you will find out how much it matters.
4. **Which two features did you add, and what question is each for?** A feature you cannot
   attach to a question is one you will never use.
5. **What is in there that should not be?** Bots, group-admin notices, one person's phone
   posting twice. You do not have to remove them yet. You have to know they are there.

---

**Where this goes next.** The features you built here are the input to everything that
follows: lesson 2 compares them across people, lesson 3 across time, lesson 4 asks what shape
they have, and lesson 5 asks which of them tell people apart. And the habit — enrich before
you model — is the one the machine learning course builds on directly.